### PR TITLE
Improve the main startup of ds2

### PR DESCRIPTION
**Pull request by sas**
This is still WIP. The goal is to make `main.cpp` a bit easier to read, make options easier to parse (by directly rejecting options that can't work in a certain mode instead of manually checking), and get closer to `lldb-server`'s command line interface to make platform mode integration easier.

### CODE REVIEW

File: Sources/main.cpp [hunk 1]
```diff
@@ -399,13 +217,12 @@ int main(int argc, char **argv) {
     env.erase(e);
   }
 
-  if (!opts.getString("attach").empty()) {
-    attachPid = atoi(opts.getString("attach").c_str());
-  }
+  int attachPid = opts.getString("attach").empty()
+                      ? -1
+                      : atoi(opts.getString("attach").c_str());
```
Comment (fjricci) [inline Sources/main.cpp]: ::atoi?

File: Sources/main.cpp [hunk 2]
```diff
@@ -435,18 +251,178 @@ int main(int argc, char **argv) {
     host = gDefaultHost;
   }
 
-  DS2LOG(Debug, "ds2 launched with arguments: %s", full_args.str().c_str());
+  std::unique_ptr<Socket> socket = CreateSocket(host, port, reverse);
+
+  if (!namedPipePath.empty()) {
+    std::string realPort = socket->port();
+    FILE *namedPipe = fopen(namedPipePath.c_str(), "a");
```
Comment (fjricci) [inline Sources/main.cpp]: ::fopen

File: Sources/main.cpp [hunk 3]
```diff
@@ -435,18 +251,178 @@ int main(int argc, char **argv) {
...
+  std::string port = server->port();
+
+  pid_t pid = ::fork();
+  if (pid < 0) {
+    DS2LOG(Fatal, "cannot fork: %s", strerror(errno));
+  }
+
+  if (pid == 0) {
+    // When in slave mode, output is suppressed but for standard error.
+    close(0);
```
Comment (fjricci) [inline Sources/main.cpp]: ::close

File: Sources/main.cpp [hunk 4]
```diff
@@ -329,8 +329,6 @@ static int PlatformMain(int argc, char **argv) {
 
     thread.detach();
   } while (true);
-
-  return EXIT_SUCCESS;
 }
```
Comment (fjricci) [inline Sources/main.cpp]: Add an unreachable annotation?

File: Sources/Utils/OptParse.cpp [hunk 5]
```diff
@@ -267,11 +267,18 @@ void OptParse::usageDie(char const *format, ...) {
   help_align += 2;
 
   for (auto const &e : _options) {
-    if (e.second.hidden)
+    if (e.second.hidden) {
```
Comment (fjricci) [inline Sources/Utils/OptParse.cpp]: I wish clang-format had a setting for this

File: Sources/Utils/OptParse.cpp [hunk 6]
```diff
@@ -267,11 +267,18 @@ void OptParse::usageDie(char const *format, ...) {
...
       continue;
+    }
 
-    Print("  -%c, --%s", e.second.shortName, e.first.c_str());
-    Print(" %s", (e.second.type == stringOption ? "ARG" : "   "));
+    static std::map<OptionType, char const *> const argTypePlaceholder = {
+        {stringOption, "ARG        "},
+        {vectorOption, "ARG        "},
```
Comment (fjricci) [inline Sources/Utils/OptParse.cpp]: Given that you're making these const-length strings (ie space-padded), it's probably easier to understand if you align them.

File: Sources/Utils/OptParse.cpp [hunk 7]
```diff
@@ -255,12 +255,22 @@ void OptParse::usageDie(char const *format, ...) {
         _parseHostPort ? " [HOST]:PORT" : "",
         _allowPositionals ? " [-- PROGRAM [ARGUMENTS...]]" : "");
 
+  static std::map<OptionType, std::string> const argTypePlaceholder = {
+      {stringOption, "ARG"},
```
Comment (fjricci) [inline Sources/Utils/OptParse.cpp]: jk looks like you deleted the spaces 😆

File: Sources/Utils/OptParse.cpp [hunk 8]
```diff
@@ -271,15 +281,11 @@ void OptParse::usageDie(char const *format, ...) {
...
-        {vectorOption, "ARG        "},
-        {hostPortOption, "[HOST]:PORT"},
-        {boolOption, "           "},
-    };
     Print("  -%c, --%s %s", e.second.shortName, e.first.c_str(),
-          argTypePlaceholder.at(e.second.type));
-    for (size_t i = 0; i < help_align - e.first.size(); ++i) {
+          argTypePlaceholder.at(e.second.type).c_str());
+    for (size_t i =
+             e.first.size() + 1 + argTypePlaceholder.at(e.second.type).size();
```
Comment (fjricci) [inline Sources/Utils/OptParse.cpp]: this is gross

File: Sources/Utils/OptParse.cpp [hunk 9]
```diff
@@ -42,22 +49,32 @@ static bool ParseHostPort(std::string const &addrString, std::string &host,
...
-  return true;
+  // Check if the host and port are understandable by `::getaddrinfo()` to make
+  // sure they are valid.
+  struct addrinfo hints, *result;
+  ::memset(&hints, 0, sizeof(hints));
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_STREAM;
+  hints.ai_flags = AI_PASSIVE;
+  hints.ai_protocol = IPPROTO_TCP;
+  return ::getaddrinfo(host.empty() ? nullptr : host.c_str(),
```
Comment (fjricci) [inline Sources/Utils/OptParse.cpp]: was going to suggest something similar ❤️

File: Sources/main.cpp (hunk 4 above)
Comment (sas) [reply Sources/main.cpp]: I don't think having an unreachable here would be of any use. These annotations are meant to help the compiler in general. Thoughts?

File: Sources/main.cpp [hunk 10]
```diff
@@ -492,7 +492,7 @@ int main(int argc, char **argv) {
     host = gDefaultHost;
   }
 
-  DS2LOG(Debug, "ds2 launched with arguments: %s", full_args.c_str());
+  DS2LOG(Debug, "ds2 launched with arguments: %s", full_args.str().c_str());
```
[the remark this answers is not in the document]
Comment (sas) [reply Sources/main.cpp]: I don't think that's correct. AIUI, this temporary object will be alive until the next sequence point.

File: Sources/main.cpp (hunk 1 above)
Comment (sas) [reply Sources/main.cpp]: There are a bunch of these from before I touched this code. I'll change all of them at once in a separate commit.

**Pull request merged**